### PR TITLE
Implementation of label smoothing with crossentropy 

### PR DESCRIPTION
**Pull request by sambitdash**
crossentropy and binarycrossentropy can have label smoothing. 

Addressing as per issue #1016.

### CODE REVIEW

File: test/layers/stateless.jl [hunk 1]
```diff
@@ -22,31 +22,42 @@ const ϵ = 1e-7
...
     @test logitcrossentropy(logŷ, y, weight = [.5, .5]) ≈ lossvalue/2
+    @test logitcrossentropy(logŷ, y, weight = ones(2), label_smoothing=0.2) ≈ lossvalue
+    @test logitcrossentropy(logŷ, y, weight = [.5, .5],label_smoothing=0.2) ≈ lossvalue/2
     @test logitcrossentropy(logŷ, y, weight = [2, .5]) ≈ 1.5049660054074199
   end
 
   logŷ, y = randn(3), rand(3)
+  ydb = [0.1, 0.9, 0.9, 0.1]
+  yb  = [1., 1., 0., 0.]
+  yr  = [2.0828626352604234, 0.3250829733914482, 2.0828626352604234, 0.3250829733914482]
```
Comment (oxinabox) [inline test/layers/stateless.jl]: Can you indicate the source of this number?

Comment (sambitdash) [reply test/layers/stateless.jl]: Hard coded numbers have been removed.

**Comment by sambitdash** [comment]
> Can we have some tests that:
> 
> 1. show label smoothing clearly differing from not using it
> 2. show properties of label soothing, e.g test that the loss with increased smoothing is larger/smaller the decreased etc

The label smoothing is a regularizer and the most significant is when the distribution of the data is skewed to one value. label smoothing ensures all zeros for weights do not end up getting to a statistically maximal category as a local minima. One a symmetric distribution over the categories the label smoothing has no effect.

**Comment by oxinabox** [comment]
> (2) is kind of very subjective of the label smoothing will reduce losses for dissimilar distributions and increases losses for similar distributions.

That sounds testable.

File: test/layers/stateless.jl [hunk 2]
```diff
@@ -25,10 +25,21 @@ const ϵ = 1e-7
   ylp = [0.9 0.1]'
   logylp = [0.0 v]'
 
+
```
Comment (oxinabox) [inline test/layers/stateless.jl]:
```suggestion
  # Construct `sim`ilar and `dis`imilar versions of the dataset so we can test effect of smoothing
  # smoothing should decrease loss on disimilar and increase the loss on similar, compared to 
  # the loss without smoothing
```

File: src/layers/stateless.jl [hunk 3]
```diff
@@ -5,56 +5,81 @@ using NNlib: logsoftmax, logσ
...
-  return -sum(y .* log.(ŷ)) .* weight * 1 // size(y, 2)
-end
+_crossentropy(ŷ::AbstractVecOrMat, y::AbstractVecOrMat, weight::Union{Number, AbstractVector}) =
+  _crossentropy(ŷ, y.*weight, nothing)
 
-function _crossentropy(ŷ::AbstractVecOrMat, y::AbstractVecOrMat, weight::AbstractVector)
-  return -sum(y .* log.(ŷ) .* weight) * 1 // size(y, 2)
+function crossentropy(ŷ::AbstractVecOrMat, y::AbstractVecOrMat; weight=nothing, label_smoothing=zero(eltype(y)))
+  if !iszero(label_smoothing)
+    y = y .* (1 - label_smoothing) .+ label_smoothing* 1 // size(y, 1)
```
Comment (oxinabox) [inline src/layers/stateless.jl]: Is it worth being explict:
```suggestion
    y = y .* (1 - label_smoothing) .+ (label_smoothing // size(y, 1))
```

Comment (sambitdash) [reply src/layers/stateless.jl]: Yes. label_smoothing can be non-integral type that the operator `//` may not support.

Comment (oxinabox) [reply src/layers/stateless.jl]: Ah..
```suggestion
    y = y .* (1 - label_smoothing) .+ (label_smoothing * 1 // size(y, 1))
```

**Comment by sambitdash** [comment]
@MikeInnes can you take a look at this change and suggest if it's good to merge.

**Comment by sambitdash** [comment]
Any reasons this is not being merged?

**Comment by CarloLucibello** [comment]
I'm not convinced that this feature is popular enough to be worth adding. Of course, I may be wrong

**Comment by oxinabox** [comment]
> I'm not convinced that this feature is popular enough to be worth adding. Of course, I may be wrong

We have a dozen less popular activation functions.
TensorFlow has this build in we should thus also.
Plus it is tricky enough to implement in a way that plays nicely with all types, that i would like a library for it.
And what other librarry but Flux?

**Comment by gxyd** [comment]
Hi, any plan to get to this in? Do we need checks to make sure that `label_smoothing` is between `0` and `1`?

I'll be interested in seeing this in.

**Comment by gxyd** [comment]
Hi, I can send in a new PR with my commits on top of this PR's commits, would that be ok?

**Comment by sambitdash** [comment]
> Hi, I can send in a new PR with my commits on top of this PR's commits, would that be ok?

@gxyd, I am no longer tracking this PR nor have the relevant set up to track this PR. I am presuming when the reviewers find the change acceptable they shall be able to merge the suitable changes. If you want to take over this PR and merge to your code, feel free to do so.

**Comment by sambitdash** [comment]
closing was completely unintentional. Reopening.

**Comment by oxinabox** [comment]
@DhairyaLGandhi this is still waiting for your review.
It's an API change

**Comment by DhairyaLGandhi** [comment]
Needs a rebase, imo better to hide the control flow in dispatch since we should check for runtimes for the backwards pass and correctness
